File: JavaScript/Medium/backtracking/subsets2.js
/**
 * @param {number []}
 * @return {number [][]}
 */

var subsets2 = function (nums) {
    let result = []
    nums.sort(function(a,b) {return a-b}) //Sorts array in ascending order by substracting a-b and using values to decide order

    var backtrack = (index, current) => {
        result.push([...current])
        for (let i = index; i<nums.length; i++){
            if (i > index && nums[i]==nums[i-1]) continue
            current.push(nums[i])
            backtrack(i+1, current)
            current.pop()
        }
    }
    backtrack(0,[])
    return result
}

File: Python/Medium/backtracking/subsets_2.py
"""
Given an integer array nums that may contain duplicates, return all possible subsets (the power set). 
"""

def subsets2(self, nums):
    result = []
    nums.sort()

    def backtrack(index, current):
        if index>=len(nums):
            if current not in nums:
                result.append(current.copy())
                return
            else:
                return
        current.append(nums[index])
        backtrack(index+1, current)

        current.pop()
        backtrack(index+1, current)
    backtrack(0,[])
    return result




